tests/topology_test.go

```diff
@@ -0,0 +1,579 @@
+	case "EdgeAdded":
+		e := msg.Obj.(*graph.Edge)
+		if g.GetEdge(e.ID) == nil {
+			g.AddEdge(e)
+		}
+	}
+
+	return nil
+}
+
+func initConfig(t *testing.T) {
+	f, err := ioutil.TempFile("", "skydive_agent")
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+
+	f.WriteString(conf)
+	f.Close()
+
+	err = config.InitConfigFromFile(f.Name())
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+}
+
+func startAgent(t *testing.T) {
+	// FIX(safchain) has to be removed see comment around the variable declaration
+	if globalAgent != nil {
+		return
+	}
+
+	initConfig(t)
+
+	err := logging.InitLogger()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	globalAgent = agent.NewAgent()
+	go globalAgent.Start()
+}
+
+func startTopologyClient(t *testing.T, g *graph.Graph, onReady func(*websocket.Conn), onChange func(*websocket.Conn)) error {
+	// ready when got a first ping
+	ws, err := connectToAgent(5, onReady)
+	if err != nil {
+		return err
+	}
+
+	for {
+		_, m, err := ws.ReadMessage()
+		if err != nil {
+			break
+		}
+
+		err = processGraphMessage(g, m)
+		if err != nil {
+			return err
+		}
+
+		logging.GetLogger().Debug("%s", string(m))
+		logging.GetLogger().Debug("%s", g.String())
+
+		onChange(ws)
+	}
+
+	return nil
+}
+
+func testTopology(t *testing.T, g *graph.Graph, cmds []string, onChange func(ws *websocket.Conn)) {
+	cmdIndex := 0
+	or := func(w *websocket.Conn) {
+		// ready to exec the first cmd
+		if cmdIndex < len(cmds) {
+			err := exec.Command("sudo", strings.Split(cmds[cmdIndex], " ")...).Run()
+			if err != nil {
+				t.Fatal(err.Error())
+			}
+			cmdIndex++
+		}
+	}
+
+	oc := func(ws *websocket.Conn) {
+		onChange(ws)
+
+		// exec the following command
+		if cmdIndex < len(cmds) {
+			err := exec.Command("sudo", strings.Split(cmds[cmdIndex], " ")...).Run()
+			if err != nil {
+				t.Fatal(err.Error())
+			}
+			cmdIndex++
+		}
+	}
+
+	err := startTopologyClient(t, g, or, oc)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+}
+
+func testCleanup(t *testing.T, g *graph.Graph, cmds []string, ints []string) {
+	// cleanup side on the test
+	testPassed := false
+	onChange := func(ws *websocket.Conn) {
+		g.Lock()
+		defer g.Unlock()
+
+		if !testPassed {
+			clean := true
+			for _, intf := range ints {
+				n := g.LookupFirstNode(graph.Metadatas{"Name": intf})
+				if n != nil {
+					clean = false
+					break
+				}
+			}
+
+			if clean {
+				testPassed = true
+
+				ws.Close()
+			}
+		}
+	}
+
+	testTopology(t, g, cmds, onChange)
+	if !testPassed {
+		t.Error("test not executed or failed")
+	}
+}
+
+func newGraph(t *testing.T) *graph.Graph {
+	backend, err := graph.NewMemoryBackend()
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+
+	g, err := graph.NewGraph(backend)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+
+	return g
+}
+
+func TestBridgeOVS(t *testing.T) {
+	g := newGraph(t)
+
+	startAgent(t)
+
+	setupCmds := []string{
+		"ovs-vsctl add-br br-test1",
+	}
+
+	tearDownCmds := []string{
+		"ovs-vsctl del-br br-test1",
+	}
+
+	testPassed := false
```

Comment:
Since these are functional tests we rely on external events like netlink, ovsdb callback and we don't know when and how the topology will be fully updated. The only thing we can say is that the topology is not the one expected after a delay. Here the maximum delay to have something corresponding to our expectation is set to 1m. So in the callback there is no real error but just incomplete informations.